### ISSUE TITLE
Standard Scaler - unit variance condition

### ISSUE THREAD

**Issue by raphaelsty** [online-ml/river#335]
Scikit Learn offers users of the standard scaler to disable the unit variance condition with the `with_std` parameter. It can happen that the standard scaler with unit variance produces values that are not conducive to model convergence (#332). 

[Standard Scaler Scikit Learn](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html)

Is this feature useful for Creme users? 

Raphaël

**Comment by etiennekintzler** [comment]
This could be useful. 

Regarding the standard scaler with unit variance,  **the magnitude of the transformed x is proportional to the frequency of change in input value**.  

Below is a plot of the transformed input value when x change from 0 to 1 at certain points for various change frequencies.

![image](https://user-images.githubusercontent.com/17943873/91464023-65d8e480-e88c-11ea-93c6-85b408c60b45.png)

As expected, less frequent changes is associated with higher transformed input. 

Also, **the transformed input does not depend on the magnitude of the change in input**. I considered changes from 0 to 1 but the result is exactly the same with change from 0 to 0.1 or from 0 to 1000. It can be shown analytically that Z=(Xi-m)/sigma does not change when we take Xi'=c*Xi.

Below is the code to reproduce the plot:

```python
# packages
import pandas as pd
import numpy  as np

import matplotlib.pyplot as plt
from sklearn import preprocessing as sk_preprocessing

# fun
def generate_scaled_values(every=100, n=1000, c=1):
    scaler = sk_preprocessing.StandardScaler()
    out = []

    for i in range(n):
        x = c if ((i+1) % every) == 0 else 0
        xarr = np.array([x]).reshape(-1, 1)

        scaler = scaler.partial_fit(xarr)
        out_tuple = (x, scaler.transform(xarr)[0][0])#, scaler_wuv.transform(xarr)[0][0])

        out += [out_tuple]

    df = pd.DataFrame(out, columns=["x", "x_transformed"])
    return df

def generate_experiment_plots(every_list=[50, 100, 200, 300, 400, 500], **kwargs):
    fig, ax = plt.subplots()
    for every in every_list:
        kwargs.update(every=every)
        df = generate_scaled_values(**kwargs)
        ax.plot(df.index, df.x_transformed, marker='o', linestyle='', ms=4, label="Every {} ({:.2f}%)".format(every, 100/every))
        ax.set_title("Transformed x for different occurences of change")
    lgd = ax.legend(bbox_to_anchor=(1, 1.025), title="Frequency of the change")

# experiment
generate_experiment_plots()
```

**Comment by MaxHalford** [comment]
Super interesting stuff. Thanks for detective work. I don't have the time to look into this at the moment. I will however add it to my todo list. It's not the end of the world, as it doesn't seem to be a bug on our side. I guess this is where we draw the line between stochastic learning, where samples are randomized, and online learning, where the data ordering matters. This is potentially impactful, and could even lead to a workshop paper or something like that. But first I want to take the time to investigate the problem myself and to see if anything has been proposed in the literature.

**Comment by etiennekintzler** [comment]
> Super interesting stuff. Thanks for the detective work. 

:smile: 

> I don't have the time to look into this at the moment. I will however add it to my todo list. It's not the end of the world, as it doesn't seem to be a bug on our side. I guess this is where we draw the line between stochastic learning, where samples are randomized, and online learning, where the data ordering matters. This is potentially impactful, and could even lead to a workshop paper or something like that. But first I want to take the time to investigate the problem myself and to see if anything has been proposed in the literature.

Sure, my goal was just to document the behavior. One solution could be to smooth out the normalization across more than multiple observations but then you'd loose the unit variance property. 

However, this solution seems out a bit of the scope for this issue. The most straightforward think to do would be to add `with_std` parameter or just use another scaler if one suspect rare occurrence of one feature !

**Comment by MaxHalford** [comment]
Closing this and adding a `with_std` parameter.

The interesting insight from this investigation is that handling features that change abruptly is not trivial. A workaround is to transform the feature manually its scale. We'll see if this merits more attention later.
